src/main/java/net/fabricmc/loader/impl/discovery/ModSolver.java

```diff
@@ -0,0 +1,980 @@
+/*
+ * Copyright 2016 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.loader.impl.discovery;
+
+import java.math.BigInteger;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+
+import org.sat4j.pb.IPBSolver;
+import org.sat4j.pb.SolverFactory;
+import org.sat4j.pb.tools.DependencyHelper;
+import org.sat4j.pb.tools.INegator;
+import org.sat4j.pb.tools.WeightedObject;
+import org.sat4j.specs.ContradictionException;
+import org.sat4j.specs.TimeoutException;
+
+import net.fabricmc.loader.api.SemanticVersion;
+import net.fabricmc.loader.api.Version;
+import net.fabricmc.loader.api.metadata.ModDependency;
+import net.fabricmc.loader.api.metadata.version.VersionBounds;
+import net.fabricmc.loader.api.metadata.version.VersionPredicate;
+import net.fabricmc.loader.impl.discovery.Explanation.ErrorKind;
+import net.fabricmc.loader.impl.util.log.Log;
+import net.fabricmc.loader.impl.util.log.LogCategory;
+import net.fabricmc.loader.impl.util.version.SemanticVersionImpl;
+
+final class ModSolver {
+	static Result solve(List<ModCandidate> allModsSorted, Map<String, List<ModCandidate>> modsById,
+			Map<String, ModCandidate> selectedMods, List<ModCandidate> uniqueSelectedMods) throws ContradictionException, TimeoutException, ModResolutionException {
+		// build priority index
+
+		Map<ModCandidate, Integer> priorities = new IdentityHashMap<>(allModsSorted.size());
+
+		for (int i = 0; i < allModsSorted.size(); i++) {
+			priorities.put(allModsSorted.get(i), i);
+		}
+
+		// create and configure solver
+
+		solverPrepTime = System.nanoTime();
+
+		IPBSolver solver = SolverFactory.newDefaultOptimizer();
+		solver.setTimeout(60); // 60s
```

Comment:
Should be <1s even for a large modpack, but doesn't hurt to leave some room. I went for 60s because it's the most I'd want to wait. It actually usually recovers itself when the limit is hit, just proceeding with a potentially non-optimal result.